paramiko/transport.py

```diff
@@ -1791,6 +2760,8 @@ def run(self):
                             self._log(ERROR, 'Channel request for unknown channel %d' % chanid)
                             self.active = False
                             self.packetizer.close()
+                        # This line is key. With it, the channel can go out of scope and shut everything down
+                        chan = None
```

Comment:
But `self._channels` also has references to the channels ... but now that I look into it, I see it uses a `WeakValueDictionary`. So that's the key here, along with your `weakref.ref` used for `transport_proxy` - "weak refs"